REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [101]:
import pandas as pd
import numpy as np

In [9]:
bus = pd.read_csv('businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('inspections.csv')

In [10]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [11]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [13]:
len(bus)

6315

In [17]:
print ("The number of records for the business.csv is", len(bus))

The number of records for the business.csv is 6315


#### Question 2b: How many unique business IDs are there?  

In [28]:
bus['business_id'].describe()

count     6315.000000
mean     48562.416627
std      33825.711496
min         10.000000
25%       5891.500000
50%      65012.000000
75%      79197.000000
max      90239.000000
Name: business_id, dtype: float64

In [29]:
bus['business_id'].unique

<bound method Series.unique of 0          10
1          19
2          24
3          31
4          45
        ...  
6310    90234
6311    90236
6312    90237
6313    90238
6314    90239
Name: business_id, Length: 6315, dtype: int64>

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [45]:
bus['city'].unique

<bound method Series.unique of 0       San Francisco
1       San Francisco
2       San Francisco
3       San Francisco
4       San Francisco
            ...      
6310    San Francisco
6311    San Francisco
6312    San Francisco
6313    San Francisco
6314    San Francisco
Name: city, Length: 6315, dtype: object>

In [61]:
bus[['name', 'city']].value_counts()

name                        city         
STARBUCKS COFFEE            San Francisco    72
PEET'S COFFEE & TEA         San Francisco    24
MCDONALDS                   San Francisco    12
WALGREENS                   San Francisco    11
SAN FRANCISCO SOUP COMPANY  San Francisco    11
                                             ..
GODIVA CHOCOLATIER          San Francisco     1
GOCHEES PIZZA               San Francisco     1
GO JOJO'S MOJO              San Francisco     1
GO GO MARKET & DELI         San Francisco     1
ZZAN                        San Francisco     1
Length: 5692, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [73]:
print(type(bus['postal_code']))

<class 'pandas.core.series.Series'>


In [81]:
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

In [86]:
print("Values of postal codes are stored as objects")

Values of postal codes are stored as objects


#### Question 3b

What are the unique values of postal_code?

In [75]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [87]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [89]:
vb = bus[bus['postal_code'].isin(validZip)]

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

In [90]:
vb['business_id'].describe()

count     5999.000000
mean     47485.367061
std      33977.547969
min         10.000000
25%       5798.000000
50%      63943.000000
75%      78780.000000
max      90239.000000
Name: business_id, dtype: float64

In [91]:
print("There are", vb['business_id'].count(), "businesses within the new parsed zip code data frame")

There are 5999 businesses within the new parsed zip code data frame


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [93]:
null_lat = vb[pd.isnull(vb['longitude'])]

In [95]:
null_lat.count()

business_id     2483
name            2483
address         2483
city            2483
state           2483
postal_code     2483
latitude           0
longitude          0
phone_number    2160
dtype: int64

In [98]:
print("There are", null_lat['business_id'].count(), "businesses with missing longitude values")

There are 2483 businesses with missing longitude values


#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [103]:
vb['null_lon'] = np.where(vb['longitude'].isnull(), 1, 0)
#create dummy variable
vb

<ipython-input-103-e4f383123c76>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vb['null_lon'] = np.where(vb['longitude'].isnull(), 1, 0)


business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  null_lon  
0     37.791116 -122.403816  +14154217044         0  
1     37.786848 -122.421547  +14157763262         0  
2     37.792888 -122.403135  +14156779494         0  
3     37.807155 -122.419004           NaN         0  
4     37.747114 -122.413641  +14156415051         0  
...         ...         ...           ...       ...  
6310        NaN         NaN           NaN         1  
6311        NaN         NaN  +14154945048         1  
6312        NaN         NaN  +14158606036         1  
6313        NaN         NaN  +14153331878         1  
6314        NaN         NaN  +14158881393         1  

[5999 rows x 10 columns]

In [104]:
vb['not_null_lon'] = np.where(vb['longitude'].isnull(), 0, 1) 

<ipython-input-104-d34a044bd083>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vb['not_null_lon'] = np.where(vb['longitude'].isnull(), 0, 1)


In [105]:
vb_df = vb[['postal_code', 'null_lon', 'not_null_lon']].groupby('postal_code').sum()
vb_df

null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.



In [108]:
vb_df['prop_null'] = vb_df['null_lon']/(vb_df['null_lon']+vb_df['not_null_lon'])
vb_df

null_lon  not_null_lon  prop_null
postal_code                                   
94102             196           262   0.427948
94103             252           306   0.451613
94104              59            74   0.443609
94105             100           126   0.442478
94107             247           204   0.547672
94108              82           142   0.366071
94109             146           237   0.381201
94110             251           334   0.429060
94111             115           167   0.407801
94112              59           131   0.310526
94114             108           122   0.469565
94115              76           162   0.319328
94116              35            58   0.376344
94117              67           130   0.340102
94118             100           143   0.411523
94121              55           106   0.341615
94122             123           147   0.455556
94123              53           116   0.313609
94124              88            76   0.536585
94127              28            43   0.394366
94131              14            34   0.291667
94132              65            69   0.485075
94133             135           282   0.323741
94134              29            45   0.391892

In [111]:
vb_df.sort_values('prop_null',ascending=False).head()

null_lon  not_null_lon  prop_null
postal_code                                   
94107             247           204   0.547672
94124              88            76   0.536585
94132              65            69   0.485075
94114             108           122   0.469565
94122             123           147   0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [112]:
ins.describe()

business_id         score          date
count  15430.000000  15430.000000  1.543000e+04
mean   40207.358004     91.179391  2.015071e+07
std    32795.456282      8.024274  8.720966e+03
min       10.000000     46.000000  2.013123e+07
25%     5215.000000     87.000000  2.014072e+07
50%    37773.500000     93.000000  2.015060e+07
75%    71057.000000     98.000000  2.016052e+07
max    89900.000000    100.000000  2.016122e+07

In [113]:
len(ins)

15430

In [114]:
pd.crosstab(ins['business_id'], columns= 'count')
# Since the count of business id's is numberous compared to the inspection data, this assumes that
# there are businesses that recieve more than one inspection

col_0        count
business_id       
10               3
19               3
24               4
31               3
45               3
...            ...
89515            1
89701            1
89790            1
89803            1
89900            1

[5730 rows x 1 columns]

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [119]:
ins['type']

0        routine
1        routine
2        routine
3        routine
4        routine
          ...   
15425    routine
15426    routine
15427    routine
15428    routine
15429    routine
Name: type, Length: 15430, dtype: object

In [120]:
pd.crosstab(ins['type'], columns= 'count') 

col_0      count
type            
complaint      1
routine    15429

In [122]:
ins2 = ins.drop(ins.loc[ins['type']=='complaint'].index) 
pd.crosstab(ins2['type'], columns= 'count')

col_0    count
type          
routine  15429

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [123]:
ins2['date'] = ins2['date'].astype(str)

In [124]:
ins2['year'] = ins2['date'].str[:4].astype('int64')
ins2

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [126]:
pd.crosstab(ins2['year'], columns= 'count') 

col_0  count
year        
2013      38
2014    5629
2015    3923
2016    5839

In [130]:
ins3 = ins2.drop(ins2.loc[ins2['year']==2013].index) 
pd.crosstab(ins3['year'], columns= 'count') 

col_0  count
year        
2014    5629
2015    3923
2016    5839

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [136]:
ins2016 = ins3.loc[ins3['year'] == 2016]
ins2016

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

In [137]:
ins2016.describe()

business_id       score    year
count   5839.000000  5839.00000  5839.0
mean   44590.954273    90.23206  2016.0
std    33997.868214     8.13340     0.0
min       10.000000    52.00000  2016.0
25%     5619.500000    86.00000  2016.0
50%    61056.000000    92.00000  2016.0
75%    77440.000000    96.00000  2016.0
max    89900.000000   100.00000  2016.0

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [138]:
bus_ins2016 = pd.merge(bus, ins2016, how='left', on=['business_id', 'business_id'])
bus_ins2016.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score      date     type    year  
0 -122.403816  +14154217044   82.0  20160503  routine  2016.0  
1 -122.421547  +14157763262   94.0  20160513  routine  2016.0  
2 -122.403135  +14156779494   98.0  20161005  routine  2016.0  
3 -122.403135  +14156779494   96.0  20160311  routine  2016.0  
4 -122.419004           NaN   84.0  20160816  routine  2016.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [141]:
bus_ins2016[['score', 'name', 'address']].sort_values('score', ascending=True).head(20)

score                         name              address
4892   52.0      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD 
7324   55.0                    POKI TIME     2101 LOMBARD ST 
6281   55.0                 CRAZY PEPPER   2257 SAN JOSE AVE 
4873   55.0                   UNCLE CAFE       65 WAVERLY PL 
3008   56.0             HING WANG BAKERY        339 JUDAH ST 
5627   56.0                   GOLDEN WOK   295 B ORIZABA AVE 
5383   57.0                   TOP'S CAFE  2460 SAN BRUNO AVE 
6067   57.0    L & G VIETNAMESE SANDWICH         602 EDDY ST 
836    57.0                 RED A BAKERY      634 CLEMENT ST 
4209   57.0      SAN TUNG RESTAURANT LLC      1031 IRVING ST 
549    57.0          MA'S DIM SUM & CAFE      1315 POWELL ST 
4903   57.0  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST 
6645   58.0    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST 
624    59.0          JOANIE'S DINER INC.   1329 COLUMBUS AVE 
6938   59.0         PAK NAM THAI CUISINE       655 LARKIN ST 
3563   59.0                  JADE GARDEN      3608 BALBOA ST 
4234   59.0                        ITHAI         720 POST ST 
6151   60.0          VALLARTA'S TACO BAR     2022 MISSION ST 
4850   60.0             HONG KONG LOUNGE     5322 GEARY BLVD 
5013   60.0                 WOLFES LUNCH        1220 16TH ST

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!

In [142]:
print("WOO HOO")

WOO HOO
